Bug 1199291 - Create result set on the fly for jobs ingested where the resultset is missing

camd opened the pull request:
This will look for the ``revision`` field in a job during ingestion.  If a resultset for that job can not be found, it will create a "skeleton" of one based on the ``revision``.  The skeleton resultset will look like:

    {
        "revision": <new_revision>,
        "push_timestamp": 0,
        "author": "pending...",
        "revisions": []
    }

Having no revisions, the resultset will not be visible in the UI until its data has been filled.  The data for the resultset will be filled by our periodic call to ``json-pushes``.

The BuildBot mechanism won't change in that we will wait to ingest jobs till we have the resultset created already.  This is for users of the ``Client`` (and the Pulse Job Ingestion, when Task Cluster starts using that).

The Client has been updated to require submitting the ``revision`` for a job and no longer requires the ``revision_hash``.  This should be a lot easier on folks submitting data.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/1356)
<!-- Reviewable:end -->

wlach commented:
Review status: 0 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[tests/model/derived/test_jobs_model.py, line 76 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV2Yg3AG89_7J3J0sK:-KDV2Yg3AG89_7J3J0sL:878655015) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L76)):*
I don't really understand the point of this structure? Why can't you use the revision data that's default associated with the sample data? Should be easy to extract a set of missing revisions from it.

---

*[tests/model/derived/test_jobs_model.py, line 93 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV2seGI0ciE6FG4KSd:-KDV2seGI0ciE6FG4KSe:-279623285) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L93)):*
Ideally we would also verify that these temporary result sets had the right sort of metadata associated with them (i.e. the "Pending..." string for author)

---

*[tests/model/derived/test_jobs_model.py, line 105 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV3BXp049NX1I6tdnv:-KDV3BXp049NX1I6tdnw:-692561798) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L105)):*
It's not clear to me why this is necessary, it doesn't look like we use "id" for anything.

---

*[tests/model/derived/test_jobs_model.py, line 115 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV3IoTxOC5ZlVx-gSz:-KDV3IoTxOC5ZlVx-gT-:-1785703728) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L115)):*
This is super similar to the previous unit test. I would use pytest.mark.parametrize to test a variety of ranges of known/unknown jobs (there's lots of examples of this in the unit test code, do a git grep for "pytest.mark.param")

---

*[tests/model/derived/test_jobs_model.py, line 179 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV3q-6tMPLHxKaervr:-KDV3q-6tMPLHxKaervs:-2022210403) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L179)):*
Some of the same issues with this test as noted above: not sure of purpose of replacing revision information, missing checks for "pending...", not sure why you're deleting the "id" property.

---

*[tests/model/derived/test_jobs_model.py, line 559 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV6zgUOGYeLOBsFoS_:-KDV6zgUOGYeLOBsFoSa:-97715441) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L559)):*
Nit: prefer r to x for variable name

---

*[treeherder/model/derived/jobs.py, line 742 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDUz5TDbu9e-XsJM7Hd:-KDUz5TDbu9e-XsJM7He:855831176) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L742)):*
Let's use `r` instead of `x`for a temporary revision.

---

*[treeherder/model/derived/jobs.py, line 758 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDUzMEvv1Ja5Zfi5O0B:-KDUzMEvv1Ja5Zfi5O0C:-1176469071) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L758)):*
I don't think there's a need to return anything here? Actually, I would probably prefer you just inline this function where it's called, as I think it'll just make things easier to follow. Your call though.

---

*[treeherder/model/derived/jobs.py, line 1808 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV4_Rqgvwfm59Am3oE:-KDV4_RrOYngSpb4vgUo:-837003200) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L1808)):*
Nit: I think I'd prefer `last_row_id`

---

*[treeherder/model/derived/jobs.py, line 1876 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV4IZVEJbSZNL-56xS:-KDV4IZVEJbSZNL-56xT:412476032) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L1876)):*
Nit: As above, I'd prefer 'r' to 'x'.

---

*[treeherder/model/derived/jobs.py, line 1881 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV7iViR8x5exBbc4Xn:-KDV7iViR8x5exBbc4Xo:-683699113) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L1881)):*
Nit, I'd prefer result_set over result

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1356)*
<!-- Sent from Reviewable.io -->

camd commented:
Review status: 0 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[tests/model/derived/test_jobs_model.py, line 76 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV2Yg3AG89_7J3J0sK:-KD_otsaWuIl1faqDYIi:-289751205) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L76)):*
yet another fair point.  I changed this to just a list comprehension to get the same struct for comparison and fetching.

---

*[tests/model/derived/test_jobs_model.py, line 93 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV2seGI0ciE6FG4KSd:-KD_WISY2pJARMzPdLvu:191446608) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L93)):*
Good point.  done.

---

*[tests/model/derived/test_jobs_model.py, line 105 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV3BXp049NX1I6tdnv:-KD_X2EqhGYgXAjYD8Kn:446958480) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L105)):*
yeah, fair enough.  I generally try to avoid comparing ids, but it is irrelevant in this case.

---

*[tests/model/derived/test_jobs_model.py, line 115 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV3IoTxOC5ZlVx-gSz:-KD_oil13c4njWUSDroJ:1346934624) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L115)):*
Yeah, good call.  I was just being lazy...  Fixed this with ``pytest.mark.parametrize``

---

*[tests/model/derived/test_jobs_model.py, line 179 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV3q-6tMPLHxKaervr:-KD_of39yLfJ2fAkQZU_:-1917171731) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L179)):*
Done.  Since this tests short converting to long, it would have been overly complex to try to parametrize, I think.  So I kept it separate, but cleaned things up a bit with your other comments in mind.

---

*[tests/model/derived/test_jobs_model.py, line 559 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV6zgUOGYeLOBsFoS_:-KDZ5wcBy_hOLbHJlMoH:-499146009) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/tests/model/derived/test_jobs_model.py#L559)):*
Done.

---

*[treeherder/model/derived/jobs.py, line 742 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDUz5TDbu9e-XsJM7Hd:-KDZ66ezG4E60yYD1mw2:-499146009) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L742)):*
Done.

---

*[treeherder/model/derived/jobs.py, line 758 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDUzMEvv1Ja5Zfi5O0B:-KDZ6tPF6gq52dmQmKgy:826495965) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L758)):*
Good call on the return, it's unnecessary...  I kind of prefer this as a separate function, if you have no strong objections.  ``store_job_data`` feels so cluttered as it is, this at least just says, "I'm doing this thing here and we can move on".  I suppose it's just a code style preference on my part.

---

*[treeherder/model/derived/jobs.py, line 1808 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV4_Rqgvwfm59Am3oE:-KDZ8cxeCp85OhsybYwe:-499146009) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L1808)):*
Done.

---

*[treeherder/model/derived/jobs.py, line 1876 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV4IZVEJbSZNL-56xS:-KDZ8ioxdaezDcN6CIYE:-499146009) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L1876)):*
Done.

---

*[treeherder/model/derived/jobs.py, line 1881 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDV7iViR8x5exBbc4Xn:-KDZ8qtQaG1eNgDlMYZh:-499146009) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L1881)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1356)*
<!-- Sent from Reviewable.io -->

wlach commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[treeherder/model/derived/jobs.py, line 758 \[r1\]](https://reviewable.io:443/reviews/mozilla/treeherder/1356#-KDUzMEvv1Ja5Zfi5O0B:-KD_rGyXJDm8ZWFSUAir:1728443497) ([raw file](https://github.com/mozilla/treeherder/blob/f494689c9f20804b0a9dea38632e900f5925510a/treeherder/model/derived/jobs.py#L758)):*
Yup, that's fine by me.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1356)*
<!-- Sent from Reviewable.io -->
